Feature/create an account

patrykmierunski opened the pull request:
Main success flow:
1. User fulfills information needed to create an account, namely: email, username and password.
3. System registers this information and hashes password
2. System informs user that account has been created and provide information about username and email address

Alternate flow 1:
1. If user with a specified email address exists, the website informs about it.

Alternate flow 2:
1. If user with a specified username exists, the website informs about it.

patrykmierunski committed:
Add lombok dependency

patrykmierunski committed:
Downgrade Spring Boot version

patrykmierunski committed:
Add @SpringBootApplication class

patrykmierunski committed:
Add springdoc-openapi-ui dependency

patrykmierunski committed:
Add temporary authentication credentials

patrykmierunski committed:
Add first working version

patrykmierunski committed:
Add password hashing

patrykmierunski committed:
Refactor naming

patrykmierunski committed:
Add data validation

Dependency and annotation were added.

patrykmierunski committed:
Refactor registration feature

greencashew commented on pom.xml:
```diff
@@ -1,64 +1,85 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
```
MINOR: I know it's project generated by spring.io, so you change formatting to your own. In case you join the team you need to ask for code formatting settings.

Cause for now it's hard to detmine for me what you changed :-)

greencashew commented on pom.xml:
```diff
@@ -1,64 +1,85 @@
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-validation</artifactId>
+        </dependency>
+        <dependency>
```
MINOR: Just from readablity perspective, you can add comment in here to separate dev environment dependencies

greencashew commented on src/main/java/pl/mantiscrab/budgetr/config/SecurityConfig.java:
```diff
@@ -0,0 +1,41 @@
+                .anyRequest().authenticated());
+        http.formLogin();
+        http.csrf().disable();
+//        http.csrf(csrf -> csrf.ignoringRequestMatchers(PathRequest.toH2Console()));
```
Remove redundant comment

greencashew commented on src/main/java/pl/mantiscrab/budgetr/config/SecurityConfig.java:
```diff
@@ -0,0 +1,41 @@
+    @Bean
+    public UserDetailsService userDetailsManager(PasswordEncoder passwordEncoder) {
+        String password = passwordEncoder.encode("password");
+        UserDetails admin = User.builder().username("admin").password(password).roles("admin").build();
```
Please explain why? And what is the further plan?

patrykmierunski replied:
Signing in will be added in next PR

greencashew commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,5 @@
+  datasource:
+    url: jdbc:h2:mem:budgetr
+    username: sa
+    password: password
```
Just keep in mind, that in normal environment it should be passed by ENVIRONMENT_VARIABLE

As far as I Remember it's redundant in here

greencashew commented on src/main/java/pl/mantiscrab/budgetr/registration/User.java:
```diff
@@ -0,0 +1,31 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true)
```
MINOR: \n between fields, it will be more readable

greencashew replied:
Still not added

greencashew commented on src/main/java/pl/mantiscrab/budgetr/registration/User.java:
```diff
@@ -0,0 +1,31 @@
+    @NotBlank
+    private String username;
+    @Column(nullable = false)
+    @NotBlank
```
Question do you really need validation in here, if you do exactly same on dto level

greencashew commented on src/main/java/pl/mantiscrab/budgetr/registration/UserMapper.java:
```diff
@@ -0,0 +1,18 @@
+import pl.mantiscrab.budgetr.registration.dto.UserDto;
+import pl.mantiscrab.budgetr.registration.dto.UserRegisterDto;
+
+class UserMapper {
```
Personally, i prefer, to make object responsible for its creation and business logic. But this is also fine, and in some cases even better ;-) 

Btw. for mapping you can also try `mapstruct` library if you wish

patrykmierunski replied:
I plan to add `mapstruct` later

greencashew commented on src/main/java/pl/mantiscrab/budgetr/registration/UserService.java:
```diff
@@ -0,0 +1,41 @@
+    private final PasswordEncoder passwordEncoder;
+    private final UserRepository userRepository;
+
+    @Transactional
```
Do you need transaction in here? :-)

greencashew commented on src/main/java/pl/mantiscrab/budgetr/registration/UserRegistrationDto.java:
```diff
@@ -0,0 +1,4 @@
+package pl.mantiscrab.budgetr.registration;
+
+record UserRegistrationDto(String email, String username, String password) {
```
For now i don't see any value with this Dto, In service layer you can operate on entity, don't be shy it's your internal

greencashew commented on src/main/java/pl/mantiscrab/budgetr/registration/UserController.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    List<ConstraintViolation> handleMethodArgumentNotValidException(MethodArgumentNotValidException ex) {
```
Good that you remembered about it :-) 

If you wish to see how I implemented that generic exception you can follow the page:
https://github.com/greencashew/spring-link-shortener/tree/main/src/main/java/dev/greencashew/link_shortener/configuration/web/exception_handling

greencashew commented on src/main/java/pl/mantiscrab/budgetr/registration/UserService.java:
```diff
@@ -0,0 +1,41 @@
+        return savedUserDto;
+    }
+
+    private void throwIfUserExist(UserRegisterDto dto) {
```
Please clearly state name, what is the purpose

patrykmierunski committed:
Changes related to Code Review comments:

-organise dependencies
-remove unnecessary comments
-remove redundant dto
-remove @Transactional
-change naming
-remove redundant datasource's user and password properties

greencashew reviewed (approved):
After discussion I'm Ok too go. Waiting for the further improvements

patrykmierunski committed:
Add integration test

patrykmierunski committed:
Add unit tests for UserService

greencashew commented on src/test/java/pl/mantiscrab/budgetr/registration/RegistrationTest.java:
```diff
@@ -0,0 +1,72 @@
+    TestRestTemplate restTemplate;
+
+    @Test
+    void registrationTest() {
```
The test method could be description

E.g.
shouldRegisterUserWhenNotExists

Or with description

greencashew commented on src/test/java/pl/mantiscrab/budgetr/registration/RegistrationTest.java:
```diff
@@ -0,0 +1,72 @@
+        assertRegisterResponseMatchesRequest(extract(registerResponse), registerRequest);
+    }
+
+    private void assertRegisterResponseMatchesRequest(UserDto registrationResponse, UserRegisterDto registerRequest) {
```
Please keep in mind that tests instead of clean should be mainly simple, and duplicates are fully fine.

For me you extracted to too many methods in this case

greencashew commented on src/test/java/pl/mantiscrab/budgetr/registration/RegistrationTest.java:
```diff
@@ -0,0 +1,72 @@
+import static pl.mantiscrab.budgetr.registration.TestUserData.sampleRegisterDto;
+
+@SpringBootTest(webEnvironment = RANDOM_PORT)
+class RegistrationTest {
```
I remember it was some way to make it functional but don't remember so far

greencashew commented on src/test/java/pl/mantiscrab/budgetr/registration/InMemoryDummyCrudRepository.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public <S extends T> S save(S entity) {
+        return null;
```
Better to throw exception if it is not implemented

greencashew commented on src/test/java/pl/mantiscrab/budgetr/registration/InMemoryDummyCrudRepository.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.Map;
+import java.util.Optional;
+
+class InMemoryDummyCrudRepository<T, ID> implements CrudRepository<T, ID> {
```
This may cause parallisation issues as you implemented in synchronous way

patrykmierunski replied:
I'm not planning to run unit tests synchronously, so I assume that this one i fine

greencashew replied:
sure just wanted to point it out

greencashew commented on src/test/java/pl/mantiscrab/budgetr/registration/InMemoryUserRepository.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    private <S extends User> User setIndexAsId(S user) {
+        return new User(index++, user.getEmail(), user.getUsername(), user.getPassword());
```
Interesting idea, I was using username as primary key 🙂

patrykmierunski committed:
Refactor tests:
-simplify RegistrationTest methods
-throw exceptions in methods not implemented in InMemoryDummyCrudRepository

greencashew commented on src/test/java/pl/mantiscrab/budgetr/registration/MethodNotImplementedException.java:
```diff
@@ -0,0 +1,4 @@
+package pl.mantiscrab.budgetr.registration;
+
+class MethodNotImplementedException extends RuntimeException{
```
You don't have to create new Exception for that, there is already one in place

patrykmierunski committed:
Remove redundant exception
